reconciletags/reconciletags.py

```diff
@@ -23,87 +23,99 @@
 
     def add_tags(self, digest, tag, dry_run):
-        logging.debug('Tagging {0} with {1}'.format(digest, tag))
-        command = ('gcloud container images add-tag {0} {1} '
-                   '-q'.format(digest, tag))
-        self.call(command, dry_run)
-
-    # This turns a list of lists into one flat list of tags
-    def flatten_tags_list(self, list_of_lists):
-        flat_tags_list = []
-        for sublist in list_of_lists:
-            for tag in sublist:
-                if tag:
-                    flat_tags_list.append(tag)
-        return flat_tags_list
-
-    def get_existing_tags(self, repo):
-        output = json.loads(self.call('gcloud container images list-tags '
-                            '--no-show-occurrences {0}'.format(repo), False))
-
-        list_of_tags = [image['tags'] for image in output]
-        existing_tags = self.flatten_tags_list(list_of_tags)
+        if not dry_run:
+            src_name = docker_name.Digest(digest)
+            dest_name = docker_name.Tag(tag)
+            creds = docker_creds.DefaultKeychain.Resolve(src_name)
+            transport = transport_pool.Http(httplib2.Http)
+
+            with docker_image.FromRegistry(
+                src_name, creds, transport) as src_img:
+                    if src_img.exists():
+                        creds = docker_creds.DefaultKeychain.Resolve(dest_name)
+                        logging.debug('Tagging {0} with {1}'.format(digest, tag))
+                        with docker_session.Push(
+                            dest_name, creds, transport) as push:
+                                push.upload(src_img)
+                    else:
+                        logging.debug("""Unable to tag {0}
+                            as the image can't be found""".format(digest))
+        else:
+            logging.debug("Would have tagged {0} with {1}".format(digest, tag))
+
+    def get_existing_tags(self, full_repo, digest):
+        full_digest = full_repo + '@sha256:' + digest
+        existing_tags = []
+
+        name = docker_name.Digest(full_digest)
+        creds = docker_creds.DefaultKeychain.Resolve(name)
+        transport = transport_pool.Http(httplib2.Http)
+
+        with docker_image.FromRegistry(name, creds, transport) as img:
+            if img.exists():
+                existing_tags = img.tags()
+            else:
+                logging.debug(
+                    """Unable to get existing tags for {0}
+                        as the image can't be found""".format(full_digest))
         return existing_tags
 
-    def get_latest_digest(self, repo):
-        output = json.loads(self.call('gcloud container images list-tags '
-                            '--no-show-occurrences {0}'.format(repo), False))
-        for image in output:
-            if 'latest' in image['tags']:
-                return image['digest']
+    def get_latest_digest(self, manifests):
+        for digest in manifests:
```

Comment:
You could return an empty string at the end, in the other case where there is no latest. That will simplify the handling below.